### ISSUE TITLE
Automated Testing for MASS

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#38]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/38) by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

Should we use this issue as a starting point for discussing automated testing in MASS?

# Short-Term #

I think a big step forward would be a set of high-level "regression tests" that simply verify the primary outputs of mass (calculations of the various p_parameters, number of pulses cut, etc). The approach would be something like the following:

* Choose a variety of pulses (pulses that both pass and fail various cuts). We don't need tens of thousands of pulses (and don't really want to have to check massive amounts of data into git).
* Run an older "trusted" version of mass against these pulses
* Save the results (primarily the `p_xxx` parameters, but probably also things like the number of pulses that pass various cuts).
* Write code to verify that the current version of mass gives the same results for all pulses.
* Provide an easy script (perhaps as part of setup.py?) to run this ... maybe even make it automatic when you do a `python setup.py build`?

Something like this could be put together quite quickly, doesn't required a special "unit testing" framework, and would provide big benefits when refactoring or performance tuning. I'm quite happy to take a first pass at this (in fact I already have some of the bits and pieces written).

# Longer Term #

It would be good to get into the habit of writing low-level automated unit tests when adding new functionality. After all, when you write new code you have to test it anyway, so you might as well make that testing automatic, right? But it isn't so simple ...

* Verifying "correct" output can be difficult. For exampe, how do you, a-priori, verify that the correct filter coefficients are being generated? If the coefficiencts can be calculated analytically for a particular pulse model, then you have a nice approach. But not all functions will be like that.
* For the types of coding that we do, it isn't always obvious what the "correct" output should be until we have worked on the code for a while. I.e., we do a lot of exploratory coding.
* Some of our coding is GUIs, which are notoriously difficult to unit test
* Our desire and need to refactor and make performance improvements means that interfaces to functions can and will change, which can break unit tests. You don't want to get into the trap of having to spend as much time maintaining unit tests as you do normal code.

I don't have any brilliant insights here, but it might be worthwhile for someone to attempt to write a good set of (low-level) unit tests for some particularly important function in mass, and then report back to everyone else on their experiences.

**Comment by joefowler** [comment]
**Original comment by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

Add regression testing to mass.

This is just a first pass at this. Much more can be done. See Issue #38
for more information.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Dan: despite all appearances, I've been thinking about this matter all autumn. 

You mention that GUIs are notoriously difficult to test, but frankly, I'm a lot more concerned about testing the numerical functions. Not that they are harder to test than GUIs, but simply that they are far more important to the success of MASS.

If we want to test numerical functions, this argues for a cleaner separation of MASS into objects and numerical code. Maybe convenience and glue code is built into the classes, but the numerical heavy lifting is done in functions. We have already made some small moves in this direction (see mass.core.analysis_algorithms, for example). Galen's thinking about a Julia-based MASS2 also points in this direction.

Still, even if the numerical functions are separated out, they are still difficult to test. How do we know what the right answers ought to be? Sometimes, simply writing the algorithm into the test, though redundant, can ensure that future bugs do not introduce changes to the source algorithm. So if we compute p_pulse_rms by subtracting the pretrigger mean and taking the rms deviation from zero, then we just write those facts into the test. For more complex algorithms, like the filter computation, we can sometimes just accept that incomplete tests are better than none. For example, we can easily test that a filter constructed to be insensitive to a baseline is exactly that, or we can test that the filter response has the expected normalization, even if we can't easily show that the noise weighting is done correctly.

Worse, some functions are random, e.g., a function meant to generate a random time series with certain correlation properties. I've done some reading about how to test random-output functions, and the consensus seems to be that you have two choices: either run the function a bazillion times and check that the results are consistent with expectations (accepting that with a five-sigma test, correct code will fail one time in a million), or replace the pseudo-random number generator with a known generator (for example, seed a generator with a fixed seed). The latter can't guarantee correctness per se, but it can guarantee that the input->output transformation is constant over time. If a person has carefully verified the output by complex, interactive means, then at least this sort of regression test can ensure that the carefully verified output doesn't change.

Our biggest issue is the question of long-term strategy on where to take Mass in the future. I agree with your comments above, and we'll make sure to implement them regardless of the future direction.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

p.s. I forgot two things in the previous comment on this testing proposal.

First, I wanted to tag @oneilg and @jojo978 so that they notice the discussion.

Second, I forgot to comment that I like your idea of a small suite of test pulses to establish that the pulse summary quantities are correct. In fact, I think we don't even need the pulses to be stored. They could be either noise-free analytic functions, or random-noisy but with a fixed seed to make them strictly repeatable. This is like what I proposed with random-output functions, except that here the functions have a fixed output (given fixed input) while the input is random. While this only tests the code to ensure that algorithms don't change, this is still valuable. I wish we'd had it going on when we attempted in October to shift some computation to Cython. Making the change without these tests was a big mistake.
